doc/PlatformDesignStory.md

```diff
@@ -0,0 +1,572 @@
+    fn call(_response: WriteCompleted) {
+        unsafe {
+            if DONE { return; }
+            DONE = true;
+        }
+        libtock_console::set_write_buffer(&NOUN);
+        libtock_console::start_write(NOUN.len());
+    }
+}
+```
+
+Now we have a reusable console library! However, we still don't have any unit
+tests.
+
+## Adding a Fake Kernel
+
+A good unit test for the console driver would test not only the driver's
+operation with successful system calls but also also test the driver's
+error-handling logic. That is difficult to do if we test using a real Tock
+kernel in an emulator -- the real kernel has the goal of avoiding system call
+errors! Instead of using a real Tock kernel, driver unit tests should use a
+"fake kernel" that simulates the kernel's functionality while allowing errors to
+be injected.
+
+To keep this document reasonably short and understandable, we have omitted error
+handling, but we can still structure our unit tests in a manner that would allow
+a test to inject errors when error handling logic is added.
+
+To allow the console driver to work with both a real kernel and a fake kernel,
+we add a `Syscalls` trait to `libtock_platform`. When compiled into a Tock
+process binary, `Syscalls` will be implemented by a zero-sized type, which
+avoids wasting non-volatile storage space or RAM area. To avoid passing around
+references to a `Syscalls` implementation, we can pass the `Syscalls` by value
+rather than by reference (i.e. take `self` rather than `&self`). For practical
+use, this requires the `Syscalls` implementations to be `Copy`.
+
+```rust
+pub trait Syscalls: Copy {
+    fn subscribe(self, driver: usize, minor: usize,
+                 callback: extern "C" fn(usize, usize, usize, usize),
+                 data: usize);
+
+    fn const_allow(self, major: usize, minor: usize, buffer: &'static [u8]);
+    fn command(self, major: usize, miner: usize, arg1: usize, arg2: usize);
+    fn yieldk(self);
+}
+```
+
+We then implement `Syscalls` using a real Tock kernel in `libtock_runtime`:
+
+```rust
+#[derive(Clone, Copy)]
+pub struct TockSyscalls;
+
+impl libtock_platform::Syscalls for TockSyscalls {
+    /* Omitted implementation details */
+}
+```
+
+We adapt `libtock_console` to use an app-provided `Syscalls` implementation
+rather than directly calling into `libtock_runtime`:
+
+```rust
+pub fn set_write_callback<S: Syscalls, Callback: FreeCallback<WriteCompleted>>(
+    syscalls: S, data: usize)
+{
+    syscalls.subscribe(1, 1, write_complete::<Callback>, data);
+}
+
+extern "C" fn write_complete<Callback: FreeCallback<WriteCompleted>>(
+    bytes: usize, _: usize, _: usize, data: usize)
+{
+    Callback::call(WriteCompleted { bytes, data });
+}
+
+pub fn set_write_buffer<S: Syscalls>(syscalls: S, buffer: &'static [u8]) {
+    syscalls.const_allow(1, 1, buffer);
+}
+
+pub fn start_write<S: Syscalls>(syscalls: S, bytes: usize) {
+    syscalls.command(1, 1, bytes, 0);
+}
```

Comment:
So all users need to create a struct `TockSyscalls` and pass that to every function?

Reply:
Yes, unfortunately. I am not aware of a way to specify a dependency that is only present in non-test builds, which is what we would need for `libtock_console` to construct the system call type itself. `TockSyscalls` is `Copy` and has no generic parameters so it should be easy to pass around.

Reply:
Since `TockSyscalls` is a ZST, it seems it would be possible to have it instantiated within the console functions that actually make the system call, with which type being used determined by `#[cfg(test)]` / `#[cfg(not(test))]`.

A playground example that illustrates what I mean: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=443aa3b33c8b65c58c532c3d222fd8da

I know that many Tock developers are pretty anti-cfg though, and this would add a lot of boilerplate code to every driver in exchange for simplifying application code by only a medium amount. Just wanted to toss it out as an idea.

Reply:
The problem with using `#[cfg(test)]` within `libtock_console` is it makes it difficult to test higher-level code. Ideally, an application crate's unit tests should be able to instantiate a copy of the application code and the console driver that use `FakeSyscalls`. However, running `cargo test` on a crate only sets `#[cfg(test)]` for that crate, not its dependencies, so if you use `#[cfg(test)]` to do the switch then `libtock_console` would end up using the real system calls rather than the fake system calls.

We could get around this by having a "fake syscalls" feature that the test code invokes. You would need to move `libtock_console`'s tests out to an external crate (and the same for every driver), but I think it would work.

Instead, I'd prefer to expose console's dependency on `Syscalls` from `libtock_console` explicitly, and re-export it with `TockSyscalls` instantiated in `libtock_core`. That is, if we move `libtock_console`'s methods into a dedicated struct (a refactoring I intended to do but left out of the doc to keep it short-ish), then we can re-export it from `libtock_core` like:

```rust
pub use libtock_console::Console<libtock_runtime::TockSyscalls> as Console;
```

Then users of `libtock_core` directly would have a version of `Console` that hides `Syscalls` entirely but at the cost of lessening testability. This seems ideal under the assumption that our users are primarily divided into two classes:

1. Users like Google that are happy to accept more boilerplate in exchange for better testability.
1. Users that want to minimize boilerplate (e.g. new users learning to use Tock and/or libtock-rs) who don't write tests for their code.

Reply:
Huh, I assumed that with a Cargo workspace it would apply the test attribute to all crates when running tests from the root, but that does not appear to be the case. 

The re-exporting technique is much cleaner anyway, I agree that seems like the best design.